File: java/var-config/src/main/java/dev/varar/config/VarConfig.java
package dev.varar.config;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * The parsed varar.config.json — the single shared config file for all var
 * tools across every language port. Same field semantics everywhere:
 * {@code docs.include} has no default (empty discovers nothing),
 * {@code docs.exclude} removes matches, both are plain globs (no {@code !}
 * prefix); {@code steps} globs step-definition files; {@code snippets} maps
 * language id to snippet template; {@code scannerPlugins} carries plugin
 * NAMES (resolution is a per-language concern — the Java port defines none
 * yet). Contract: conformance/config/README.md. All keys optional; unknown
 * keys, wrong types, and malformed JSON fail loudly (a typo'd config must
 * never silently discover nothing); a {@code $schema} key is ignored.
 */
public record VarConfig(
        List<String> docsInclude,
        List<String> docsExclude,
        List<String> steps,
        Map<String, String> snippets,
        List<String> scannerPlugins) {

    private static final Set<String> KNOWN_KEYS = Set.of("$schema", "docs", "steps", "snippets", "scannerPlugins");
    private static final Set<String> KNOWN_DOCS_KEYS = Set.of("include", "exclude");

    public VarConfig {
        docsInclude = List.copyOf(docsInclude);
        docsExclude = List.copyOf(docsExclude);
        steps = List.copyOf(steps);
        snippets = Map.copyOf(snippets);
        scannerPlugins = List.copyOf(scannerPlugins);
    }

    public static VarConfig empty() {
        return new VarConfig(List.of(), List.of(), List.of(), Map.of(), List.of());
    }

    /** Reads {@code <root>/varar.config.json}; a missing file is the empty config. */
    public static VarConfig load(Path root) {
        Path path = root.resolve("varar.config.json");
        if (!Files.isRegularFile(path)) return empty();
        String text;
        try {
            text = Files.readString(path, StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new IllegalArgumentException(path + ": " + e.getMessage(), e);
        }
        return parse(text, path.toString());
    }

    /** Pure parse of the config TEXT; {@code sourceName} prefixes every error message. */
    public static VarConfig parse(String jsonText, String sourceName) {
        Object data;
        try {
            data = Json.parse(jsonText);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(sourceName + ": invalid JSON: " + e.getMessage(), e);
        }
        if (!(data instanceof Map<?, ?> map)) {
            throw new IllegalArgumentException(sourceName + ": top level must be an object");
        }
        Set<String> unknown = new TreeSet<>();
        for (Object key : map.keySet()) {
            if (!KNOWN_KEYS.contains((String) key)) unknown.add((String) key);
        }
        if (!unknown.isEmpty()) {
            throw new IllegalArgumentException(sourceName + ": unknown key(s): " + String.join(", ", unknown));
        }
        List<String> docsInclude = List.of();
        List<String> docsExclude = List.of();
        Object docs = map.get("docs");
        if (docs != null) {
            if (!(docs instanceof Map<?, ?> docsMap)) {
                throw new IllegalArgumentException(sourceName + ": 'docs' must be an object");
            }
            Set<String> unknownDocs = new TreeSet<>();
            for (Object key : docsMap.keySet()) {
                if (!KNOWN_DOCS_KEYS.contains((String) key)) unknownDocs.add((String) key);
            }
            if (!unknownDocs.isEmpty()) {
                throw new IllegalArgumentException(
                        sourceName + ": unknown docs key(s): " + String.join(", ", unknownDocs));
            }
            docsInclude = stringList(docsMap.get("include"), "docs.include", sourceName);
            docsExclude = stringList(docsMap.get("exclude"), "docs.exclude", sourceName);
        }
        Map<String, String> snippets = new LinkedHashMap<>();
        Object rawSnippets = map.get("snippets");
        if (rawSnippets != null) {
            if (!(rawSnippets instanceof Map<?, ?> snippetsMap)) {
                throw new IllegalArgumentException(sourceName + ": 'snippets' must be an object of strings");
            }
            for (Map.Entry<?, ?> entry : snippetsMap.entrySet()) {
                if (!(entry.getValue() instanceof String value)) {
                    throw new IllegalArgumentException(sourceName + ": 'snippets' must be an object of strings");
                }
                snippets.put((String) entry.getKey(), value);
            }
        }
        return new VarConfig(
                docsInclude,
                docsExclude,
                stringList(map.get("steps"), "steps", sourceName),
                snippets,
                stringList(map.get("scannerPlugins"), "scannerPlugins", sourceName));
    }

    private static List<String> stringList(Object value, String key, String sourceName) {
        if (value == null) return List.of();
        if (!(value instanceof List<?> list)) {
            throw new IllegalArgumentException(sourceName + ": '" + key + "' must be an array of strings");
        }
        List<String> out = new ArrayList<>(list.size());
        for (Object item : list) {
            if (!(item instanceof String s)) {
                throw new IllegalArgumentException(sourceName + ": '" + key + "' must be an array of strings");
            }
            out.add(s);
        }
        return List.copyOf(out);
    }
}

File: dotnet/Varar.Config/VarConfig.cs
using System.Collections.Immutable;
using System.Text.Json;
using Varar.Core;

namespace Varar.Config;

/// <summary>Thrown on a malformed <c>varar.config.json</c>. The message starts with the config path.</summary>
public sealed class VarConfigException : Exception
{
    public VarConfigException(string message)
        : base(message)
    {
    }
}

/// <summary>Spec-doc discovery globs. Both are plain globs (no <c>!</c> prefix).</summary>
public sealed record VarGlobs(ImmutableArray<string> Include, ImmutableArray<string> Exclude);

/// <summary>
/// The parsed, unresolved shape of <c>varar.config.json</c> — pure data, shared byte-for-byte with
/// every port's reader. Port of <c>config-types.ts</c>.
/// </summary>
public sealed record ParsedVarConfig(
    VarGlobs Docs,
    ImmutableArray<string> Steps,
    ImmutableDictionary<string, string> Snippets);

/// <summary>
/// Strict, fail-loud reader of <c>varar.config.json</c>. Port of <c>config.ts</c>: no defaults for
/// docs/steps (a repo declares both), unknown keys and wrong types are errors, missing file → empty.
/// </summary>
public static class VarConfig
{
    private static readonly HashSet<string> KnownKeys =
        new(StringComparer.Ordinal) { "$schema", "docs", "steps", "snippets" };

    private static readonly HashSet<string> KnownDocsKeys =
        new(StringComparer.Ordinal) { "include", "exclude" };

    public static readonly ParsedVarConfig Empty = new(
        new VarGlobs([], []),
        [],
        ImmutableDictionary<string, string>.Empty);

    /// <summary>Parses config text (no filesystem). Fails loudly with the path prefixed.</summary>
    public static ParsedVarConfig Parse(string jsonText, string sourcePath)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(jsonText);
        }
        catch (JsonException e)
        {
            throw new VarConfigException($"{sourcePath}: invalid JSON: {e.Message}");
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                throw new VarConfigException($"{sourcePath}: top level must be an object");
            }

            foreach (var prop in root.EnumerateObject())
            {
                if (!KnownKeys.Contains(prop.Name))
                {
                    throw new VarConfigException(
                        $"{sourcePath}: unknown key \"{prop.Name}\" (known keys: docs, steps, snippets)");
                }
            }

            var docs = Empty.Docs;
            if (root.TryGetProperty("docs", out var docsEl) && docsEl.ValueKind != JsonValueKind.Null)
            {
                if (docsEl.ValueKind != JsonValueKind.Object)
                {
                    throw new VarConfigException($"{sourcePath}: \"docs\" must be an object");
                }

                foreach (var prop in docsEl.EnumerateObject())
                {
                    if (!KnownDocsKeys.Contains(prop.Name))
                    {
                        throw new VarConfigException($"{sourcePath}: unknown key \"docs.{prop.Name}\" (known: include, exclude)");
                    }
                }

                docs = new VarGlobs(
                    StringArray(docsEl, "include", "docs.include", sourcePath),
                    StringArray(docsEl, "exclude", "docs.exclude", sourcePath));
            }

            var snippets = ImmutableDictionary<string, string>.Empty;
            if (root.TryGetProperty("snippets", out var snippetsEl) && snippetsEl.ValueKind != JsonValueKind.Null)
            {
                if (snippetsEl.ValueKind != JsonValueKind.Object ||
                    snippetsEl.EnumerateObject().Any(x => x.Value.ValueKind != JsonValueKind.String))
                {
                    throw new VarConfigException($"{sourcePath}: \"snippets\" must be an object of strings");
                }

                var builder = ImmutableDictionary.CreateBuilder<string, string>();
                foreach (var entry in snippetsEl.EnumerateObject())
                {
                    builder[entry.Name] = entry.Value.GetString()!;
                }

                snippets = builder.ToImmutable();
            }

            return new ParsedVarConfig(
                docs,
                StringArray(root, "steps", "steps", sourcePath),
                snippets);
        }
    }

    /// <summary>Reads <c>varar.config.json</c> from <paramref name="cwd"/>, or the empty config if absent.</summary>
    public static ParsedVarConfig Load(string cwd)
    {
        var path = Path.Combine(cwd, "varar.config.json");
        return File.Exists(path) ? Parse(File.ReadAllText(path), path) : Empty;
    }

    /// <summary>Projects the parsed config to the shared conformance wire shape.</summary>
    public static Value ToArtifact(ParsedVarConfig config) => Value.Map([
        new("docs", Value.Map([
            new("include", Value.List(config.Docs.Include.Select(Value.Of))),
            new("exclude", Value.List(config.Docs.Exclude.Select(Value.Of))),
        ])),
        new("steps", Value.List(config.Steps.Select(Value.Of))),
        new("snippets", Value.Map(config.Snippets.Select(kv => new KeyValuePair<string, Value>(kv.Key, Value.Of(kv.Value))))),
    ]);

    private static ImmutableArray<string> StringArray(JsonElement parent, string key, string label, string sourcePath)
    {
        if (!parent.TryGetProperty(key, out var value) || value.ValueKind == JsonValueKind.Null)
        {
            return [];
        }

        if (value.ValueKind != JsonValueKind.Array ||
            value.EnumerateArray().Any(v => v.ValueKind != JsonValueKind.String))
        {
            throw new VarConfigException($"{sourcePath}: \"{label}\" must be an array of strings");
        }

        return [.. value.EnumerateArray().Select(v => v.GetString()!)];
    }
}
